plugins/inputs/elasticsearch_query/elasticsearch_query.go

```diff
@@ -0,0 +1,277 @@
+		}
+		e.httpclient = httpclient
+	}
+
+	clientOptions = append(clientOptions,
+		elastic.SetHttpClient(e.httpclient),
+		elastic.SetSniff(e.EnableSniffer),
+		elastic.SetURL(e.URLs...),
+		elastic.SetHealthcheckInterval(e.HealthCheckInterval.Duration),
+	)
+
+	if e.Username != "" && e.Password != "" {
+		clientOptions = append(clientOptions,
+			elastic.SetBasicAuth(e.Username, e.Password),
+		)
+	}
+
+	if e.HealthCheckInterval.Duration == 0 {
+		clientOptions = append(clientOptions,
+			elastic.SetHealthcheck(false),
+		)
+	}
+
+	if e.Tracelog {
+		clientOptions = append(clientOptions,
+			elastic.SetTraceLog(log.New(os.Stdout, "", log.LstdFlags)),
+		)
+	}
+
+	client, err := elastic.NewClient(clientOptions...)
+	if err != nil {
+		return err
+	}
+
+	// check for ES version on first node
+	esVersion, err := client.ElasticsearchVersion(e.URLs[0])
+
+	if err != nil {
+		return fmt.Errorf("Elasticsearch query version check failed: %s", err)
+	}
+
+	// quit if ES version is not supported
+	i, err := strconv.Atoi(strings.Split(esVersion, ".")[0])
+	if err != nil || i < 5 {
+		return fmt.Errorf("Elasticsearch query: ES version not supported: %s", esVersion)
+	}
+
+	e.ESClient = client
+	return nil
+}
+
+func (e *ElasticsearchQuery) Gather(acc telegraf.Accumulator) error {
+	if err := e.init(); err != nil {
+		return err
+	}
+
+	e.acc = acc
+
+	var wg sync.WaitGroup
+
+	for _, agg := range e.Aggregations {
+		wg.Add(1)
+		go func(agg Aggregation) {
+			defer wg.Done()
+			err := e.esAggregationQuery(agg)
+			if err != nil {
+				acc.AddError(fmt.Errorf("Elasticsearch query aggregation %s: %s ", agg.MeasurementName, err.Error()))
+			}
+		}(agg)
+	}
+
+	wg.Wait()
+	return nil
+}
+
+func (e *ElasticsearchQuery) createHttpClient() (*http.Client, error) {
+	tlsCfg, err := e.ClientConfig.TLSConfig()
+	if err != nil {
+		return nil, err
+	}
+	tr := &http.Transport{
+		ResponseHeaderTimeout: e.Timeout.Duration,
+		TLSClientConfig:       tlsCfg,
+	}
+	httpclient := &http.Client{
+		Transport: tr,
+		Timeout:   e.Timeout.Duration,
+	}
+
+	return httpclient, nil
+}
+
+func (e *ElasticsearchQuery) esAggregationQuery(aggregation Aggregation) error {
+
+	ctx, cancel := context.WithTimeout(context.Background(), e.Timeout.Duration)
+	defer cancel()
+
+	if aggregation.DateField == "" {
+		return fmt.Errorf("Field 'date_field' not set")
+	}
+
+	mapMetricFields, err := e.getMetricFields(ctx, aggregation)
+	if err != nil {
+		return err
+	}
+
+	aggregationQueryList, err := e.buildAggregationQuery(mapMetricFields, aggregation)
+	if err != nil {
+		return err
+	}
+
+	searchResult, err := e.runAggregationQuery(ctx, aggregation, aggregationQueryList)
+	if err != nil {
+		return err
+	}
+
+	if searchResult.Aggregations != nil {
+		err = e.parseAggregationResult(&aggregationQueryList, searchResult)
+		if err != nil {
+			return err
+		}
+	} else {
+		err = e.parseSimpleResult(aggregation.MeasurementName, searchResult)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
```

Comment:
You can shorten to
```suggestion
	if searchResult.Aggregations != nil {
		err = e.parseAggregationResult(&aggregationQueryList, searchResult)
	} else {
		err = e.parseSimpleResult(aggregation.MeasurementName, searchResult)
	}

	return err
```
But that's a matter of taste...